File: Test Driven Development/kata/tdd-bowling-kata/scala/src/main/scala/BowlingGame.scala
class BowlingGame {

  var rolls: List[Int] = List()

  def roll(pins: Int) { rolls = rolls :+ pins }

  def score: Int = (convert(rolls) take 20).sum

  def convert(list: List[Int]): List[Int] = {
    list match {
      case 10 :: x :: y :: tail => 10 :: (x + y) :: convert(x :: y :: tail)
      case x :: y :: z :: tail if (x + y == 10) => 10 :: z :: convert(z :: tail)
      case x :: tail => x :: convert(tail)
      case nil => Nil
    }
  }
}

File: Test Driven Development/kata/tdd-bowling-kata/java/src/main/java/kata/BowlingGame.java
package kata;

public class BowlingGame {

    private int rolls[] = new int[21];
    private int currentRoll = 0;

    public void roll(int pins) {
        rolls[currentRoll++] = pins;
    }

    public int score() {
        int score = 0;
        int frameIndex = 0;
        for (int frames = 0; frames < 10; frames++) {
            if (isStrike(frameIndex)) {
                score += strikeBonus(frameIndex);
                frameIndex++;
            }
            else if (isSpire(frameIndex)) {
                score += spireBonus(frameIndex);
                frameIndex += 2;
            } else {
                score += sumOfBallsInFrame(frameIndex);
                frameIndex += 2;
            }
        }
        return score;
    }

    private boolean isStrike(int frameIndex) {
        return rolls[frameIndex] == 10;
    }

    private int sumOfBallsInFrame(int frameIndex) {
        return rolls[frameIndex] + rolls[frameIndex + 1];
    }

    private int spireBonus(int frameIndex) {
        return 10 + rolls[frameIndex + 2];
    }

    private int strikeBonus(int frameIndex) {
        return 10 + rolls[frameIndex + 1] + rolls[frameIndex + 2];
    }

    private boolean isSpire(int frameIndex) {
        return rolls[frameIndex] + rolls[frameIndex + 1] == 10;
    }
}


